File: swift/MovingAverageFromDataStream.swift
// https://leetcode.com/problems/moving-average-from-data-stream/

class MovingAverage {
    
    private var size: Int = 0
    private var sum: Int = 0
    private var arr = [Int]()

    init(_ size: Int) {
        self.size = size
    }
    
    func next(_ val: Int) -> Double {
        if arr.count == size {
            sum -= arr.removeFirst()
        }
        sum += val
        arr.append(val)
        
        return Double(sum) / Double(arr.count)
    }
}

File: c/MovingAverageFromDataStream.c
// https://leetcode.com/problems/moving-average-from-data-stream/

typedef struct {
    int *array;
    int capacity, size, headIndex;
    double sum;
} MovingAverage;

MovingAverage* movingAverageCreate(int size) {
    MovingAverage *obj = (MovingAverage *)malloc(sizeof(MovingAverage));
    obj->array = (int *)malloc(sizeof(int) * size);
    obj->capacity = size;
    obj->size = 0;
    obj->headIndex = -1;
    obj->sum = 0.0;
    return obj;
}

double movingAverageNext(MovingAverage* obj, int val) {
    obj->headIndex = (obj->headIndex + 1) % obj->capacity;
    if (obj->size >= obj->capacity) {
        obj->sum -= obj->array[obj->headIndex];
    }
    obj->sum += obj->array[obj->headIndex] = val;
    obj->size++;
    return obj->sum / (obj->size > obj->capacity ? obj->capacity : obj->size);
}

void movingAverageFree(MovingAverage* obj) {
    free(obj->array);
    free(obj);
}